apps/backend/src/modules/plan/controller.ts

```diff
@@ -0,0 +1,290 @@
+
+  const gt = await PlanModel.findOne({ userEmail: context.user.email });
+  if (!gt) {
+    throw new Error("No Plan found for this user");
+  }
+  const tmp = formatPlan(gt);
+  return tmp ? [tmp] : [];
+}
+
+// delete a planTerm specified by ObjectID
+export async function removePlanTerm(
+  planTermID: string,
+  context: any
+): Promise<string> {
+  if (!context.user.email) throw new Error("Unauthorized");
+
+  // check if planTerm belongs to plan
+  const gt = await PlanModel.findOne({ userEmail: context.user.email });
+  if (!gt) {
+    throw new Error("No Plan found for this user");
+  }
+  const planTermIndex = gt.planTerms.findIndex(
+    (sem) => (sem._id as string) == planTermID
+  );
+  if (planTermIndex === -1) {
+    throw new Error("PlanTerm does not exist in user's plan");
+  }
+  gt.planTerms.splice(planTermIndex, 1);
+  console.log(planTermIndex, gt.planTerms);
+  await gt.save();
+
+  return planTermID;
+}
+
+// create a new planTerm
+export async function createPlanTerm(
+  mainPlanTerm: PlanTermInput,
+  context: any
+): Promise<PlanTerm> {
+  if (!context.user.email) throw new Error("Unauthorized");
+  const nonNullPlanTerm = omitBy(mainPlanTerm, (value) => value == null);
+  nonNullPlanTerm.userEmail = context.user.email;
+  const newPlanTerm = new PlanTermModel({
+    ...nonNullPlanTerm,
+  });
+
+  // add to plan in chronological order
+  const gt = await PlanModel.findOne({ userEmail: context.user.email });
+  if (!gt) {
+    throw new Error("No Plan found for this user");
+  }
+  gt.planTerms.push(newPlanTerm);
+  await gt.save();
+
+  return formatPlanTerm(newPlanTerm);
+}
+
+// update an existing planTerm
+export async function editPlanTerm(
+  planTermID: string,
+  mainPlanTerm: EditPlanTermInput,
+  context: any
+): Promise<PlanTerm> {
+  if (!context.user.email) throw new Error("Unauthorized");
+  const gt = await PlanModel.findOne({ userEmail: context.user.email });
+  if (!gt) {
+    throw new Error("No Plan found for this user");
+  }
+
+  const planTermIndex = gt.planTerms.findIndex(
+    (sem) => (sem._id as string) == planTermID
+  );
+  if (planTermIndex === -1) {
+    throw new Error("PlanTerm does not exist in user's plan");
+  }
+
+  const termToUpdate = gt.planTerms[planTermIndex];
+
+  if (mainPlanTerm.name != null) {
+    termToUpdate.name = mainPlanTerm.name;
+  }
+  if (mainPlanTerm.year != null) {
+    termToUpdate.year = mainPlanTerm.year;
+  }
+  if (mainPlanTerm.term != null) {
+    termToUpdate.term = mainPlanTerm.term;
+  }
+  if (mainPlanTerm.hidden != null) {
+    termToUpdate.hidden = mainPlanTerm.hidden;
+  }
+  if (mainPlanTerm.status != null) {
+    termToUpdate.status = mainPlanTerm.status;
+  }
+  if (mainPlanTerm.pinned != null) {
+    termToUpdate.pinned = mainPlanTerm.pinned;
+  }
+  if (mainPlanTerm.courses != null) {
+    termToUpdate.courses = mainPlanTerm.courses.map(
+      (courseInput) => new SelectedCourseModel(courseInput)
+    );
+  }
```

Comment:
Turn into for loop on keys instead of this? This is a bit unnecessarily verbose, unless mainPlanTerm has a lot of other properties that we do not want this logic for.